arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/RaiseAccumulate.kt

```diff
@@ -951,65 +968,206 @@ public open class RaiseAccumulate<Error>(
+    newErrors.toNonEmptyListOrNull()?.let(::accumulateAll)
+  }
+
+  @OptIn(ExperimentalRaiseAccumulateApi::class)
+  @PublishedApi
+  @Deprecated("Binary compatibility", level = DeprecationLevel.HIDDEN)
+  internal fun hasErrors(): Boolean = hasAccumulatedErrors
+
+  @Suppress("KotlinUnreachableCode") // wrong inspection
+  @OptIn(ExperimentalRaiseAccumulateApi::class)
+  @PublishedApi
+  @Deprecated("Binary compatibility", level = DeprecationLevel.WARNING)
+  internal fun raiseErrors(): Nothing = latestError?.value ?: error("No accumulated errors to raise")
+
+  @Suppress("NOTHING_TO_INLINE")
+  @Deprecated(message = "Deprecated in favor of member", level = DeprecationLevel.HIDDEN)
+  public inline operator fun <A> Value<A>.getValue(thisRef: Nothing?, property: KProperty<*>): A = value
+
+  public sealed class Value<out A> {
+    public abstract val value: A
+
+    @Suppress("NOTHING_TO_INLINE")
+    public inline operator fun getValue(thisRef: Nothing?, property: KProperty<*>): A = value
+  }
+
+  @PublishedApi
+  internal class Error(private val raise: () -> Nothing) : Value<Nothing>() {
+    @OptIn(ExperimentalRaiseAccumulateApi::class)
+    @Deprecated("Binary compatibility", level = DeprecationLevel.HIDDEN)
+    constructor(raiseAccumulate: RaiseAccumulate<*>) : this({
+      raiseAccumulate.latestError?.value ?: error("No accumulated errors to raise")
+    })
+    // WARNING: do not turn this into a property with initializer!!
+    //          'raiseErrors' is then executed eagerly, and leads to wrong behavior!!
+    override val value get(): Nothing = raise()
+  }
+
+  @PublishedApi internal class Ok<out A>(override val value: A): Value<A>()
 
-  @ExperimentalRaiseAccumulateApi
-  public fun <A> Either<Error, A>.bindOrAccumulate(): Value<A> =
-    accumulating { this@bindOrAccumulate.bind() }
 
   @ExperimentalRaiseAccumulateApi
-  public fun <A> Iterable<Either<Error, A>>.bindAllOrAccumulate(): Value<List<A>> =
-    accumulating { this@bindAllOrAccumulate.bindAll() }
+  public inline fun <A> accumulating(block: RaiseAccumulate<Error>.() -> A): Value<A> {
+    contract { callsInPlace(block, AT_MOST_ONCE) }
+    return merge {
+      Ok(block(RaiseAccumulate(this@RaiseAccumulate) { raise(accumulate(it)) }))
+    }
+  }
 
   @ExperimentalRaiseAccumulateApi
-  public fun <A> EitherNel<Error, A>.bindNelOrAccumulate(): Value<A> =
-    accumulating { this@bindNelOrAccumulate.bindNel() }
+  @RaiseDSL
+  public inline fun <A> recover(
+    block: RaiseAccumulate<Error>.() -> A,
+    recover: (error: Error) -> A,
+  ): A {
+    contract { callsInPlace(block, AT_MOST_ONCE) }
+    return arrow.core.raise.recover({
+      block(RaiseAccumulate(this@RaiseAccumulate, ::raise))
+    }, recover)
+  }
 
   @ExperimentalRaiseAccumulateApi
-  public inline fun ensureOrAccumulate(condition: Boolean, raise: () -> Error) {
+  public inline fun ensureOrAccumulate(condition: Boolean, raise: () -> Error): Value<Unit> {
```

Comment:
Changed the return type here to allow `by` delegation on this method like its other siblings. Note that it's experimental, so such a breaking change is fine.